# generateMutation 0.1.9024
* config experiment fixed coverage & boosted resources

# generateMutation 0.1.9023
* config experiment genologin

# generateMutation 0.1.9022
* mutated and base reads

# generateMutation 0.1.9021
* dag.minimal

# generateMutation 0.1.9020
* config.dev

# generateMutation 0.1.9019
* scaling config and ressources

# generateMutation 0.1.9018
* simplified folders for connection with detectMutations

# generateMutation 0.1.9017
* snakemake 5.25.0 on genologin to handle allow_missing=True

# generateMutation 0.1.9016
* genologin.yaml updated

# generateMutation 0.1.9015
* obtained expected AF

# generateMutation 0.1.9014
* fixed Chr names for snps and mutated
* full working v2
* AF and reads origin to be fixed

# generateMutation 0.1.9013
* new version to debug (numerous changes)
* config: new names and list for parameters to explore
* config.dag: reduced config for dag and tests
* cp_files: copy files
* generate_mutations: faster
* iss_base & iss_mutated: generate reads
* merge_reads: simplified
* samtools_faidx: sample reference
* vcf2model & simuG: heterozigosity

# generateMutation 0.1.9012
* fix generate_mutations.R

# generateMutation 0.1.9011
* fix chromosome name

# generateMutation 0.1.9010
* multiple chromosome version

# generateMutation 0.1.9009
* data/
* updated README

# generateMutation 0.1.9008
* fixed job.sh (python conflict) and get_source

# generateMutation 0.1.9007
* integrated Biostrings with singularity on my GitHub

# generateMutation 0.1.9006
* 1st full pipeline
* generate_mutations
* generate_reads
* merge_reads
* rename index & uncompress

# generateMutation 0.1.9005
* new organisation and rules

# generateMutation 0.1.9004
* runcompress reference
* generate reads
* split reads
* mutate.py

# generateMutation 0.1.9003
* renaming rules
* config/genologin.yaml
* uncompress reference
* generate reads

# generateMutation 0.1.9002
* reorganize
* prepare genologin test

# generateMutation 0.1.9001
* subsample
