---
title: "generate Mutations"
author: Sylvain Schmitt
date: April 20, 2021
output:
  github_document: 
    toc: yes
---

```{r setup, include=FALSE}
rm(list = ls()) ; invisible(gc()) ; set.seed(42)
library(knitr)
library(kableExtra)
if(knitr:::is_html_output()) options(knitr.table.format = "html") 
if(knitr:::is_latex_output()) options(knitr.table.format = "latex") 
library(tidyverse)
theme_set(bayesplot::theme_default())
opts_chunk$set(echo = F, message = F, warning = F, fig.height = 6, fig.width = 8,
               cache = F, cache.lazy = F)
```

Development of a [`singularity` & `snakemake`](https://github.com/sylvainschmitt/snakemake_singularity) workflow to generate *in silico* mutations.

# Source

Get source data.

* Rule: [`get_source`](https://github.com/sylvainschmitt/generateMutations/blob/main/rules/get_source.smk), [`uncompress_source`](https://github.com/sylvainschmitt/generateMutations/blob/main/rules/uncompress_source.smk) & [`index_source`](https://github.com/sylvainschmitt/generateMutations/blob/main/rules/index_source.smk)
* Tools: `snakemake.remote.HTTP`, `mv`, `zcat` & [`samtools faidx`](http://www.htslib.org/doc/samtools-faidx.html)
* Address: http://`r yaml::read_yaml("config/config.yml")$genome$address`
* Files: `r yaml::read_yaml("config/config.yml")$genome$sequence`, `r yaml::read_yaml("config/config.yml")$genome$genes`, `r yaml::read_yaml("config/config.yml")$genome$TE`
* Singularity: "oras://registry.forgemia.inra.fr/gafl/singularity/samtools/samtools:latest"

# **Base** reference

Subsample genome with chromosome and positions.

* Rules: [`reference_bed`](https://github.com/sylvainschmitt/generateMutations/blob/main/rules/reference_bed.smk) & [`sample_reference`](https://github.com/sylvainschmitt/generateMutations/blob/main/rules/sample_reference.smk)
* Tools: `echo` & [`bedtools getfasta`](https://bedtools.readthedocs.io/en/latest/content/tools/getfasta.html)
* Singularity: oras://registry.forgemia.inra.fr/gafl/singularity/bedtools/bedtools:latest
* Sample: `r yaml::read_yaml("config/config.yml")$reference$chromosome`:`r yaml::read_yaml("config/config.yml")$reference$start`-`r yaml::read_yaml("config/config.yml")$reference$stop`

# **Mutated** reference

Mutate reference.

* Rules: [`generate_mutations`](https://github.com/sylvainschmitt/generateMutations/blob/main/rules/generate_mutations.smk)
* Script: [`generate_mutations.R`](https://bedtools.readthedocs.io/en/latest/content/scripts/generate_mutations.R)
* Singularity: https://depot.galaxyproject.org/singularity/bioconductor-biostrings:2.58.0--r40hd029910_1
* Mutations:
    * Number: `r yaml::read_yaml("config/config.yml")$mutations$number`
    * Transition/Transversion ratio (see below): `r yaml::read_yaml("config/config.yml")$mutations$R`

To be upgraded on several chromosomes and with annotation (beware R might start to be slow for such a task): transposable elements, heterosigozity, genes.

```{r R}
include_graphics("https://dridk.me/images/post17/transition_transversion.png")
```

# **Base** reads

Generate reads from base reference.

* Rules:
* Tools: [`insilicoseq`](https://insilicoseq.readthedocs.io/en/latest/)
* Singularity:  "docker://hadrieng/insilicoseq:latest"
* Parameters:
    * Sequencing machine: `r yaml::read_yaml("config/config.yml")$sequencing$illumina` 
    * Number of reads: `r yaml::read_yaml("config/config.yml")$sequencing$n_reads` 

# **Mutated** reads

Generate mixed reads from base and mutated reference.

* Rules:
* Tools: [`insilicoseq`](https://insilicoseq.readthedocs.io/en/latest/)
* Singularity:  "docker://hadrieng/insilicoseq:latest"
* Parameters:
    * Allelic fraction: `r yaml::read_yaml("config/config.yml")$mutations$AF` 
    * Sequencing machine: `r yaml::read_yaml("config/config.yml")$sequencing$illumina` 
    * Number of reads: `r yaml::read_yaml("config/config.yml")$sequencing$n_reads` 

# Misc

## Commands

```{bash, eval=F, echo=T}
snakemake -np 
snakemake --dag | dot -Tsvg > dag/dag.svg
snakemake --use-singularity
snakemake --report results/report.html
```

## DAG

```{r}
knitr::include_graphics("dag/dag.svg")
```

## Benchamrk

```{r benchmark}
files <- list.files("results/benchmarks", full.names = T)
names(files) <- gsub(".benchmark.txt", "", gsub("results/benchmarks/", "", files))
lapply(files, read_tsv, na = c("-", "NA")) %>% 
  bind_rows(.id = "rule") %>% 
  ggplot(aes(reorder(rule, s), `h:m:s`)) +
  geom_col() +
  coord_flip() +
  xlab("") + ylab("Time")
```

## Resources

* [TreeMutation pages](https://treemutation.netlify.app/mutations-detection.html#in-silico-mutations)
* [genologin skanemake template](https://forgemia.inra.fr/bios4biol/workflows/-/tree/06c6a5cb3206a594f9a535ba8d3df3e64682a8bc/Snakemake/template_dev)
* [Oak genome A4 snakemake](https://forgemia.inra.fr/genome_a4/genome_a4)
* [singularity images from forgemia](https://forgemia.inra.fr/gafl/singularity)
* [biocontainers](https://biocontainers.pro/tools/bioconductor-biostrings)

### GitHub

* https://github.com/ShixiangWang/sigminer
* https://github.com/ShixiangWang/sigflow
* https://github.com/FunGeST/Palimpsest
* https://github.com/IARCbioinfo/needlestack
* https://github.com/luntergroup/octopus
* https://github.com/G3viz/g3viz

## References
