---
title: "generate Mutations"
author: Sylvain Schmitt
date: April 20, 2021
output:
  github_document: 
    toc: yes
---

```{r setup, include=FALSE}
rm(list = ls()) ; invisible(gc()) ; set.seed(42)
library(knitr)
library(kableExtra)
if(knitr:::is_html_output()) options(knitr.table.format = "html") 
if(knitr:::is_latex_output()) options(knitr.table.format = "latex") 
library(tidyverse)
theme_set(bayesplot::theme_default())
opts_chunk$set(echo = F, message = F, warning = F, fig.height = 6, fig.width = 8,
               cache = F, cache.lazy = F)
```

Development of a [`singularity` & `snakemake`](https://github.com/sylvainschmitt/snakemake_singularity) workflow to generate *in silico* mutations.

# Summary

# Data

## Get genome

From `r yaml::read_yaml("config/config.yml")$genome` using `wget`.

## Chromosmes length

Using `bioawk`.

```{r refStats}
read_tsv("results/data/chromosomes_length.txt", col_names = c("name", "length")) %>% 
  filter(grepl("Chr", name)) %>% 
  mutate(sel = ifelse(name == yaml::read_yaml("config/config.yml")$chromosome, 1, 0)) %>% 
  ggplot(aes(name, length/10^6, fill = as.factor(sel))) + 
  geom_col() +
  coord_flip() +
  xlab("Chromosome") +
  ylab("Mb") +
  scale_fill_manual("Selected", values = c("grey", "darkblue"))
```

## Sample reference

Using `seqkit gep`. [Help](https://bioinf.shenwei.me/seqkit/usage/#grep).

# Reads

## Uncompress reference

For `insilicoseq` using `gzip`.

## Generate reads

Using `insilicoseq`. [Help](https://insilicoseq.readthedocs.io/en/latest/iss/generate.html#full-list-of-options).

## Split reads

Using `seqkit split2` and `mv`. [Help](https://bioinf.shenwei.me/seqkit/usage/#split2).

# Mutations

## Generate mutations

How?

* Using `SomaticSpike`.
* Using script `mutate.py` on the reference and generating again reads to be mixde for mutation frequency
* Using [`msbar` from `EMBOSS`](http://emboss.sourceforge.net/apps/release/6.4/emboss/apps/msbar.html) with previous strategy

# All

## Commands

```{bash, eval=F, echo=T}
snakemake -np 
snakemake --dag | dot -Tsvg > dag/dag.svg
snakemake --use-singularity
snakemake --report results/report.html
```

## DAG

```{r}
knitr::include_graphics("dag/dag.svg")
```

## Benchamrk

```{r benchmark}
files <- list.files("results/benchmarks", full.names = T)
names(files) <- gsub(".benchmark.txt", "", gsub("results/benchmarks/", "", files))
lapply(files, read_tsv) %>% 
  bind_rows(.id = "rule") %>% 
  ggplot(aes(reorder(rule, s), `h:m:s`)) +
  geom_col() +
  coord_flip() +
  xlab("") + ylab("Time")
```

# Resources

* [TreeMutation pages](https://treemutation.netlify.app/mutations-detection.html#in-silico-mutations)
* [genologin skanemake template](https://forgemia.inra.fr/bios4biol/workflows/-/tree/06c6a5cb3206a594f9a535ba8d3df3e64682a8bc/Snakemake/template_dev)
* [Oak genome A4 snakemake](https://forgemia.inra.fr/genome_a4/genome_a4)
* [singularity images from forgemia](https://forgemia.inra.fr/gafl/singularity)

## GitHub

* https://github.com/ShixiangWang/sigminer
* https://github.com/ShixiangWang/sigflow
* https://github.com/FunGeST/Palimpsest
* https://github.com/IARCbioinfo/needlestack
* https://github.com/luntergroup/octopus
* https://github.com/G3viz/g3viz

# References
